From 95a960dd97b07e10f7da127f4dbe39aac5f1d2a7 Mon Sep 17 00:00:00 2001
From: Ewan Milne <emilne@redhat.com>
Date: Wed, 20 Aug 2014 18:59:22 -0400
Subject: [scsi] scsi_debug: fix logical block provisioning support when unmap_alignment != 0

Message-id: <1408561176-18133-44-git-send-email-emilne@redhat.com>
Patchwork-id: 88469
O-Subject: [RHEL7.1 PATCH e-stor 43/57] [SCSI] scsi_debug: fix logical block provisioning support when unmap_alignment != 0
Bugzilla: 1132100
RH-Acked-by: David Milburn <dmilburn@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: "Ewan D. Milne" <emilne@redhat.com>

upstream commit a027b5b90eb6b5ec5c5890a5e4e198f458ae94cb

Tested by Ewan D. Milne and Barry Donahue with a combination
of manual and automated testing.  No issues seen.

upstream commit a027b5b90eb6b5ec5c5890a5e4e198f458ae94cb
Author: Akinobu Mita <akinobu.mita@gmail.com>
Date:   Mon Aug 26 22:08:41 2013 +0900

    [SCSI] scsi_debug: fix logical block provisioning support when unmap_alignment != 0

    Commit b90ebc3d5c41c9164ae04efd2e4f8204c2a186f1 ("[SCSI] scsi_debug:
    fix logical block provisioning support") fixed several issues with
    logical block provisioning support, but it still doesn't properly fix
    the cases when unmap_alignment > 0.

    For example, load scsi_debug module with the following module parameters
    and make all blocks mapped by filling the storage with zero.

            # modprobe scsi_debug lbpu=1 unmap_alignment=1 unmap_granularity=4
            # dd if=/dev/zero of=$DEV

    Then, try to unmap the first unmappable blocks at lba=1, but GET LBA STATUS
    unexpectedly reports that the last UNMAP has done nothing.

            # sg_unmap --lba=1 --num=4 $DEV
            # sg_get_lba_status --lba=1 $DEV
            descriptor LBA: 0x0000000000000001  blocks: 16383  mapped

    The problem is in map_index_to_lba(), which should return the first
    LBA which is corresponding to a given index of provisioning map
    (map_storep).

    Signed-off-by: Akinobu Mita <akinobu.mita@gmail.com>
    Acked-by: "Martin K. Petersen" <martin.petersen@oracle.com>
    Acked-by: Douglas Gilbert <dgilbert@interlog.com>
    Signed-off-by: James Bottomley <JBottomley@Parallels.com>

Signed-off-by: Ewan D. Milne <emilne@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/scsi/scsi_debug.c b/drivers/scsi/scsi_debug.c
index 2f39b13..01c0ffa 100644
--- a/drivers/scsi/scsi_debug.c
+++ b/drivers/scsi/scsi_debug.c
@@ -1997,8 +1997,14 @@ static unsigned long lba_to_map_index(sector_t lba)
 
 static sector_t map_index_to_lba(unsigned long index)
 {
- return index * scsi_debug_unmap_granularity -
-  scsi_debug_unmap_alignment;
+ sector_t lba = index * scsi_debug_unmap_granularity;
+
+ if (scsi_debug_unmap_alignment) {
+  lba -= scsi_debug_unmap_granularity -
+   scsi_debug_unmap_alignment;
+ }
+
+ return lba;
 }
 
 static unsigned int map_state(sector_t lba, unsigned int *num)
-- 
1.7.1